cpp/src/arrow/util/hashing.h

```diff
@@ -511,7 +512,9 @@ struct SmallScalarTraits<Scalar,
 template <typename Scalar, template <class> class HashTableTemplateType = HashTable>
 class SmallScalarMemoTable : public MemoTable {
  public:
-  explicit SmallScalarMemoTable(MemoryPool* pool, int64_t entries = 0) {
+  explicit SmallScalarMemoTable(MemoryPool* pool, int64_t entries = 0)
+    : index_to_value_(std::vector<Scalar, stl_allocator<Scalar>>(stl_allocator<Scalar>(pool)))
```

Comment:
What's the point of doing this? By construction this is a small private container.

Reply:
The idea here is to allocate vector `index_to_value_`  on supplied memory pool rather than alloc it directly. This is done because according our testing (with valgrind massif tool) vector  `index_to_value_` requires non-negligeble amount of memory and it is much better to allocate it on the memory pool rather than directly (since the usage of memory pool can be tracked/monitored by the user).

Reply:
Why non-negligible? It should accumulate at most 256 elements, and it's a temporary vector. What are your findings exactly?

Reply:
The idea is that by allocating these few STL strucures on the memory pool (instead of allocating directly) massif said that now only ~1% is allocated outside the pool compared to ~90% which was previously (the number however depends on many things like dictionary encoding, etc.). I don't remember the exact breakdown between these structures since I did this experiment a while ago, but I will look for the exact outputs soon.

Reply:
Are you sure you're benchmarking the right thing? As @pitrou pointed, this instance is capped at 256 elements. The allocator isn't going to change anything here since it's copied afterward (in a proper Buffer from a memory pool).

Reply:
@fsaintjacques where does the copy over happen it seems like first this is populated and kept around for some amount of time?  I could see an argument if N (where is large) then this would take non-negligible off-memory pool memory, but I'm not sure if there are any scenarios where that occurs.

Reply:
I'm removing these changes.